src/librustc_typeck/check/wfcheck.rs

```diff
@@ -354,19 +343,100 @@ impl<'a, 'gcx> CheckTypeWellFormedVisitor<'a, 'gcx> {
                 }
             }
 
-            let predicates = fcx.tcx.predicates_of(item_def_id).instantiate_identity(fcx.tcx);
-            let predicates = fcx.normalize_associated_types_in(item.span, &predicates);
-            this.check_where_clauses(fcx, item.span, &predicates);
+            this.check_where_clauses(fcx, item.span, item_def_id);
 
             fcx.impl_implied_bounds(item_def_id, item.span)
         });
     }
 
+    /// Checks where clauses and inline bounds that are declared on def_id.
     fn check_where_clauses<'fcx, 'tcx>(&mut self,
                                        fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
                                        span: Span,
-                                       predicates: &ty::InstantiatedPredicates<'tcx>)
+                                       def_id: DefId) {
+        self.inner_check_where_clauses(fcx, span, def_id, false)
+    }
+
+    fn check_trait_where_clauses<'fcx, 'tcx>(&mut self,
+                                       fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
+                                       span: Span,
+                                       def_id: DefId) {
+        self.inner_check_where_clauses(fcx, span, def_id, true)
+    }
+
+    /// Checks where clauses and inline bounds that are declared on def_id.
+    fn inner_check_where_clauses<'fcx, 'tcx>(&mut self,
+                                       fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
+                                       span: Span,
+                                       def_id: DefId,
+                                       is_trait: bool)
     {
+        use ty::subst::Subst;
+        use rustc::ty::TypeFoldable;
+
+        let mut predicates = fcx.tcx.predicates_of(def_id);
+        let mut substituted_predicates = Vec::new();
+
+        let generics = self.tcx.generics_of(def_id);
+        let defaulted_params = generics.types.iter()
+                                             .filter(|def| def.has_default &&
+                                                     def.index >= generics.parent_count() as u32);
+        for param_def in defaulted_params {
+            // Defaults must be well-formed.
+            let d = param_def.def_id;
+            fcx.register_wf_obligation(fcx.tcx.type_of(d), fcx.tcx.def_span(d), self.code.clone());
+            // Check the clauses are well-formed when the param is substituted by it's default.
+            // In trait definitions, the predicate `Self: Trait` is problematic.
+            // Therefore we skip such predicates. This means we check less than we could.
```

Comment:
Here too I'd like to see an example.

---

As an example, consider something like this:

```rust
struct Foo<T = u32, U = i32>
where T: Bar<U> { }

trait Bar<U> where Self: Baz<U> { }

trait Baz<U> { }
impl Baz<String> for u32 { }
impl Baz<i32> for String { }
```

Here, when processing `Foo`, we would attempt to prove the following:

```
forall<T, U> { // i.e., with T, U in scope
    (T: Bar<U>) => { // and with `T: Bar<U>` in our environment
        WF(T: Bar<U>), 
        WF(u32: Bar<U>), // substituting the default for T
        WF(T: Baz<i32>),
    }
}
```

This would fail because

- `WF(u32: Bar<U>)` requires that `u32: Baz<U>`, which is not implemented.
- `WF(T: Bar<i32>)` requires that `T: Baz<i32>`, which is not implemented.

---

Actually, before I go too much further, let me verify again to be sure I am remembering correctly... that *is* the intention of this PR, right?